From: Tony Espy <espy@canonical.com>
Date: Wed, 19 Mar 2014 16:57:05 -0400
Subject: src: allow MessageCenter/Proxy for TYPE_INTERNET

Allow a MessageCenter and MessageProxy to be set for an
OFONO_GPRS_CONTEXT_TYPE_INTERNET contexts in order to
support combined Internet/MMS contexts.

[ratchanan@ubports.com: forward-port to oFono 1.29]
---
 ofono/src/gprs.c | 19 +++++++++++++------
 1 file changed, 13 insertions(+), 6 deletions(-)

diff --git a/ofono/src/gprs.c b/ofono/src/gprs.c
index ed78947..f1c2719 100644
--- a/ofono/src/gprs.c
+++ b/ofono/src/gprs.c
@@ -3,6 +3,7 @@
  *  oFono - Open Source Telephony
  *
  *  Copyright (C) 2008-2011  Intel Corporation. All rights reserved.
+ *  Copyright (C) 2014  Canonical Ltd.
  *  Copyright (C) 2015-2021  Jolla Ltd.
  *
  *  This program is free software; you can redistribute it and/or modify
@@ -805,9 +806,10 @@ static void pri_reset_context_settings(struct pri_context *ctx)
 
 	pri_context_signal_settings(ctx, signal_ipv4, signal_ipv6);
 
-	if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_MMS) {
+	if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_MMS)
 		pri_set_ipv4_addr(interface, NULL);
 
+	if (ctx->proxy_host != NULL) {
 		g_free(ctx->proxy_host);
 		ctx->proxy_host = NULL;
 		ctx->proxy_port = 0;
@@ -1089,7 +1091,7 @@ static void append_context_properties(struct pri_context *ctx,
 	ofono_dbus_dict_append(dict, "AuthenticationMethod", DBUS_TYPE_STRING,
 				&strvalue);
 
-	if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_MMS) {
+	if (ctx->message_center != NULL) {
 		strvalue = ctx->message_proxy;
 		ofono_dbus_dict_append(dict, "MessageProxy",
 					DBUS_TYPE_STRING, &strvalue);
@@ -1750,7 +1752,8 @@ static DBusMessage *pri_set_property(DBusConnection *conn,
 		return pri_set_auth_method(ctx, conn, msg, str);
 	}
 
-	if (ctx->type != OFONO_GPRS_CONTEXT_TYPE_MMS)
+	if (ctx->type != OFONO_GPRS_CONTEXT_TYPE_MMS ||
+		ctx->type != OFONO_GPRS_CONTEXT_TYPE_INTERNET)
 		return __ofono_error_invalid_args(msg);
 
 	if (!strcmp(property, "MessageProxy")) {
@@ -2332,7 +2335,7 @@ static void write_context_settings(struct ofono_gprs *gprs,
 	g_key_file_set_boolean(gprs->settings, context->key, "Preferred",
 				context->preferred);
 
-	if (context->type == OFONO_GPRS_CONTEXT_TYPE_MMS) {
+	if (context->message_center != NULL) {
 		g_key_file_set_string(gprs->settings, context->key,
 					"MessageProxy",
 					context->message_proxy);
@@ -2847,7 +2850,9 @@ static void provision_context(const struct ofono_gprs_provision_data *ap,
 	strcpy(context->context.apn, ap->apn);
 	context->context.proto = ap->proto;
 
-	if (ap->type == OFONO_GPRS_CONTEXT_TYPE_MMS) {
+	if (ap->type == OFONO_GPRS_CONTEXT_TYPE_MMS ||
+		ap->type == OFONO_GPRS_CONTEXT_TYPE_INTERNET) {
+
 		if (ap->message_proxy != NULL)
 			strcpy(context->message_proxy, ap->message_proxy);
 
@@ -3779,7 +3784,9 @@ static gboolean load_context(struct ofono_gprs *gprs, const char *group)
 	if (apn == NULL)
 		goto error;
 
-	if (type == OFONO_GPRS_CONTEXT_TYPE_MMS) {
+	if (type == OFONO_GPRS_CONTEXT_TYPE_MMS ||
+		type == OFONO_GPRS_CONTEXT_TYPE_INTERNET) {
+
 		msgproxy = g_key_file_get_string(gprs->settings, group,
 						"MessageProxy", NULL);
 
